File: linked list/reverse_k_nodes_linkedlist.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */

let reversetillstop = function(head, stop){
    let tmp = head, prevnode = null;
    while(tmp != stop){
        let nextnode = tmp.next;
        tmp.next = prevnode;
        prevnode = tmp;
        tmp = nextnode;
    }
    return prevnode;
}

var reverseKGroup = function(head, k) {
    let steps = 0;
    let stop = head;
    while(stop != null && steps < k){
        stop = stop.next;
        steps++;
    }
    
    if(steps < k) return head;

    let newhead = reversetillstop(head, stop);
    let tailofreversed = head;//post reversal new tail
    let nextstart = stop;//new starting point is the last stop point
    //connect new tail to new head but new head needs to be computed first so recursive call 
    tailofreversed.next = reverseKGroup(nextstart, k);

    return newhead;
};

File: linked list/reverse_k_nodes_linkedlist.cpp
/**
 * Definition for singly-linked list.
 * class Node {
 * public:
 *     int data;
 *     Node *next;
 *     Node() : data(0), next(nullptr) {}
 *     Node(int x) : data(x), next(nullptr) {}
 *     Node(int x, Node *next) : data(x), next(next) {}
 * };
 */

 Node* reverse(Node *head){
    Node* tmp = head;
    Node *prevnode = nullptr;
    while(tmp != nullptr){
        Node *nextnode = tmp->next;
        tmp->next = prevnode;
        prevnode = tmp;
        tmp = nextnode;
    }
    return prevnode;
}

Node* kReverse(Node* head, int k) {
    Node* curstart = head, *lastnewtail = nullptr, *finalhead = nullptr;
    while(curstart != nullptr){
        //find kth node
        Node* kthnode = curstart;
        int cnt = 0;
        while(kthnode != nullptr){
            cnt++;
            if(cnt == k) break;
            kthnode = kthnode->next;
        }

        //store the pointer to next segment if exists
        Node* nextstart = (kthnode != nullptr)? kthnode->next : nullptr;
        Node* newhead;

        if(cnt < k){
            //kth node is invalid for this case
            //no reversal needed
            newhead = curstart;
            if(lastnewtail != nullptr) lastnewtail->next = curstart;
        }else{
            //disconnect k sized LL before reversal
            kthnode->next = nullptr;
            newhead = reverse(curstart);
            //connect previous segment if exists(for first segment its not) to current
            if(lastnewtail != nullptr) lastnewtail->next = newhead;
            //update lastnewtail
            lastnewtail = curstart;
        }

        //just runs once sole purpose to find final head in final answer
        if(curstart == head) finalhead = newhead;
        curstart = nextstart;
    }

    return finalhead;
}